rfc-116-store-attachment-data-in-content-items.md

```diff
@@ -0,0 +1,147 @@
+# Store attachment data in content items
+
+## Summary
+
+Add a new field to the details hash of content items, `attachments`,
+which has metadata about the page's attachments (if any).
+
+
+## Problem
+
+Pages on GOV.UK can have attachments, which come in a few different
+types.  Whitehall, which probably has the richest model of
+attachments, has:
+
+- External attachments (links to other websites)
+- HTML attachments
+- File attachments (which can have previews)
+
+Attachments are referenced in the content of the page, and make up
+part of the govspeak (or HTML) the publishing app sends to the
+publishing-api.
+
+Attachments are *not* referenced anywhere else in the content item.
+To get the details of an attachment, you have to parse the body of the
+page.  This restricts what we can do with attachments.  For example:
+
+- We cannot generate comprehensive [schema.org][] metadata for attachments: [see this comment](https://github.com/alphagov/govuk_publishing_components/pull/1247#pullrequestreview-338008254).
+- Publishing API is unable to tell Asset Manager to take an asset out of draft state, which means publishing apps have to communicate with both.
+
+Additionally, users of the content API cannot make use of attachments
+without parsing the page body.  This inhibits creative use of our
+content.
+
+
+## Proposal
+
+1. Add a new required field called `attachments` (of type [`asset_link_list`][]) to the `details` of formats which can have attachments.
+
+2. Add a new optional `preview_url` field to the `asset_link` type, making the schema:
+
+   ```
+   {
+     asset_link: {
+       type: "object",
+       additionalProperties: false,
+       required: [
+         "url",
+         "content_type",
+       ],
+       properties: {
+         content_id: {
+           "$ref": "#/definitions/guid",
+         },
+         url: {
+           type: "string",
+           format: "uri",
+         },
+         preview_url: {
+           type: "string",
+           format: "uri",
+         },
+         content_type: {
+           type: "string",
```

Comment:
What do you envisage `content_type` to be? Are we talking `content_type: "image/png"` or something domain specific such as `content_type: "HtmlAttachment"? In which case, would each content type conform to its own schema?

Reply:
`image/png`, that sort of thing.  See https://www.gov.uk/api/content/employment-tribunal-decisions/mr-f-clayton-v-royal-mail-group-ltd-1800801-2017 for an example